### PR TITLE
Simplify API basis

### PR DESCRIPTION
**Pull request by BalzaniEdoardo**
## API Simplifications
This PR is intended as a step towards simplifying the definition of a custom basis. In particular the following simplifications to the API have been carried on:

- removed the `mode` attribute in basis.
- re-computed input shape at every `compute_features` call, by always invoking `setup_basis`
- removed the abstract method `_set_input_independent_states`. This is called internally by the setup_basis of our NeMoS implementation but won't be needed for a valid custom basis.
- `_evaluate` moved to public method. This is used to compute the kernels and by `evaluate_on_grid`.
- rely on function to get equi-spaced samples in `_get_samples`. This would make it easier to generalize the behavior of `evaluate_on_grid` for a custom basis. Note that evaluate on grid may not make sense for a general basis. For example, if the basis is handling videos, i.e. the input is very high dimensional, then creating a grid in such high dimension is unfeasible and the visualization does not make sense for that. We should probably implement a try/catch if the basis do not implement `evaluate_on_grid`, then raise a `NotImplementedError` and it is up to a user to define the method if needed. 
- Infer input dimensionality from `compute_features` signature if private attribute is not available.
- `set_kernel` method of ConvBasisMixin made private.

#### Initialization of Composite basis
- Moved input dimensionality computation to `CompositeBasis`. 
- Infer input dim based on `compute_features` if attribute `_n_input_dimensionality` is not found.

#### Abstract Methods
- `setup_basis`: compute states like the input shape and kernels for convolution.
- `set_input_shape`: sets the input shape without requiring the `compute_features` call.
- `evaluate`: performs the evaluation of the basis at provided sample points. it is the non-linearity itself, it is used by `evaluate_on_grid` and `setup_basis` in `ConvBasisMixin` bases to compute the kernels. Apply the non-linearity to all elements of a ND array. does not reshape the input, add an extra dimension of size `n_basis_funcs`.
- `_compute_features`: calls evaluate and reshape to 2D by flattening all the dimensions but the first. This private method is an "unsafe" `compute_features` that does not call `setup_basis` nor perform checks. `CompositeBasisMixin` rely on this method to avoid setting up the sub-bases multiple times. If in the future we will want to guarantee that the basis are auto-differentiable, this is the method we should target.

#### Public Concrete Methods
- `compute_features`: calls `setup_basis`, then `_check_input_shape_consistency`, finally  `_compute_features`.
- `evaluate_on_grid`: gets the samples. As a default relies on the function `get_equi_spaced_samples` but behavior can be overwritten by re-implementing a `_get_samples` method.
- `split_by_feature`: splits the an array over its feature axis into a bunch of slices, one for each additive component. For our basis, it further reshape the output to  `(*input.shape[1:], n_basis_funcs)`. For a custom basis, the best we will be able to do is split by component (since we will could store the output shape once compute feature is called), but we won't know a priori how to reshape the feature axis.
- `get_params`/`set_params`/`to_transformer`: sklearn compatibility methods.
- `set_kernel`: this is for `ConvBasisMixin` and should be probably dropped, requiring only a `setup_basis` for simplicity.

#### Improved Private methods
- `_check_input_dimensionality` won't rely on the attribute `_n_input_dimensionality`, and try to infer the number of time series input based on `compute_features` if not provided.

### Unexpected failure of doctests 
Adding more examples for the now public method evaluate changed the state of the numpy seed, this resulted in divergent recurrent simulations. I enforced a seed and reduced the coefficient magnitude to assure stability.

### CODE REVIEW

**Review by billbrod — CHANGES_REQUESTED** [review]
This looks pretty good to me, just a couple of small clarifying questions:
- TransformerBasis still requires set_input_shape to be explicitly called, but that should be avoidable at this point, right? (as discussed in #305)
- The nesting of methods is getting confusing for me: `compute_features` is instantiated by the concrete classes, but they just call `super().compute_features`, which is defined in `Basis` (it's instantiated in the concrete classes so we can add the right docstring?), which calls `setup_basis` (defined in `EvalBasisMixin` and `ConvBasisMixin`) and _check_input_shape_consistency (defined in `AtomicBasisMixin` / `CompositeBasisMixin`) and `_compute_features` (defined in `EvalBasisMixin` / `ConvBasisMixin`), which in turn calls `evaluate` (defined in the concrete classes, but they're just calling `super().evaluate`, defined in the non-eval/conv version of the concrete classes). is that right?

File: src/nemos/_inspect_utils/inspect_utils.py [hunk 1]
```diff
@@ -245,3 +245,22 @@ def trim_kwargs(cls: type, kwargs: dict, class_specific_params: dict):
         for key, value in kwargs.items()
         if key in class_specific_params[cls.__name__]
     }
+
+
+def count_params_by_kind(func: Callable, kind: set[inspect.Parameter]):
+    """Count how many parameters of the callable are of the desired kind."""
```
[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply src/nemos/_inspect_utils/inspect_utils.py:251]: I used kind since that's how it is called by inspect, `parameter.kind` is the attribute storing the flag indicating what type of argument it is. I changed the type annotation to be more precise, and I added a note explain the parameter kind in the docstrings.

File: src/nemos/basis/_composition_utils.py [hunk 2]
```diff
@@ -209,3 +211,35 @@ def _atomic_basis_label_setter_logic(
         )
         bas._label = new_label
     return
+
+
+def _check_input_consistency(bas, *x):
```
[the remark this answers is not in the document]
Comment (BalzaniEdoardo) [reply src/nemos/basis/_composition_utils.py]: Removed together with the corresponding method

File: docs/how_to_guide/plot_05_transformer_basis.md [hunk 3]
```diff
@@ -110,18 +126,15 @@ trans_bas = (composite_basis).to_transformer()
...
-try:
-    trans_bas.fit_transform(inp)
-except RuntimeError as e:
-    print(repr(e))
+trans_bas.fit_transform(inp)
     
 ```
 
-...Unfortunately, not yet. The problem is that the basis doesn't know which columns of `inp` should be processed by `count_basis` and which by `speed_basis`.
+### Defining the Input Shape
```
Comment (billbrod) [inline docs/how_to_guide/plot_05_transformer_basis.md:133]:
```suggestion
### Defining the Input Shape

:::{note}
If each basis accepts a single column of the input, as in [this example](sklearn-how-to), then the default behavior works fine, and `set_input_shape` does not need to be called.
:::
```
